cpp/src_prims/metrics/randIndex.h

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * @file randIndex.h
+ * @todo TODO(Ganesh Venkataramana):
+ * <pre>
+ * The below randIndex calculation implementation is a Brute force one that uses (nElements*nElements) threads (2 dimensional grids and blocks)
+ * For small datasets, this will suffice; but for larger ones, work done by the threads increase dramatically. 
+ * A more mathematically intensive implementation that uses half the above threads can be done, which will prove to be more efficient for larger datasets
+ * the idea is as follows:
+  * instead of 2D block and grid configuration with a total of (nElements*nElements) threads (where each (i,j) through these threads represent an ordered pair selection of 2 data points),
+  a 1D block and grid configuration with a total of (nElements*(nElements))/2 threads (each thread index represents an element part of the set of unordered pairwise selections from the dataset (nChoose2))
+  * In this setup, one has to generate a one-to-one mapping between this 1D thread index (for each kernel) and the unordered pair of chosen datapoints.
+  * More specifically, thread0-> {dataPoint1, dataPoint0}, thread1-> {dataPoint2, dataPoint0}, thread2-> {dataPoint2, dataPoint1} ... thread((nElements*(nElements))/2 - 1)-> {dataPoint(nElements-1),dataPoint(nElements-2)}
+  * say ,
+     * threadNum: thread index | threadNum = threadIdx.x + BlockIdx.x*BlockDim.x,
+     * i : index of dataPoint i
+     * j : index of dataPoint j
+  * then the mapping is as follows:
+     * i = ceil((-1 + sqrt(1 + 8*(1 + threadNum)))/2) = floor((1 + sqrt(1 + 8*threadNum))/2)
+     * j = threadNum - i(i-1)/2
+  * after obtaining the the pair of datapoints, calculation of rand index is the same as done in this implementation
+ * Caveat: since the kernel implementation involves use of emulated sqrt() operations:
+  * the number of instructions executed per kernel is ~40-50 times
+  * as the O(nElements*nElements) increase beyond the floating point limit, floating point inaccuracies occur, and hence the above floor(...) !=  ceil(...)
+ * </pre>
+ */
+#include "cuda_utils.h"
+#include <thrust/reduce.h>
+#include <thrust/device_ptr.h>
+#include <cub/cub.cuh>
+#include <math.h>
+#include "common/device_buffer.hpp"
+#include "common/cuml_allocator.hpp"
+
+
+namespace MLCommon{
+namespace Metrics{
+
+
+/**
```

Comment:
remove this backup method. Not required anymore in light of its faster version below.